dartagnan/src/main/java/com/dat3m/dartagnan/program/event/arch/riscv/AmoAbstract.java

```diff
@@ -0,0 +1,80 @@
+package com.dat3m.dartagnan.program.event.arch.riscv;
+
+import com.dat3m.dartagnan.expression.IExpr;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.Tag.RISCV;
+import com.dat3m.dartagnan.program.event.core.MemEvent;
+import com.dat3m.dartagnan.program.event.core.utils.RegReaderData;
+import com.dat3m.dartagnan.program.event.core.utils.RegWriter;
+import com.dat3m.dartagnan.program.event.visitors.EventVisitor;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+
+import static com.dat3m.dartagnan.program.event.Tag.*;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.sosy_lab.java_smt.api.Formula;
+
+public abstract class AmoAbstract extends MemEvent implements RegWriter, RegReaderData {
+
+    protected final Register resultRegister;
+    protected final Register r2;
+    protected Formula memLoadValueExpr;
+    protected Formula memStoreValueExpr;
+
+    AmoAbstract(Register rd, Register r2, IExpr address, String mo) {
+        super(address, mo);
+        this.resultRegister = rd;
+        this.r2 = r2;
+        addFilters(ANY, VISIBLE, MEMORY, READ, WRITE, REG_WRITER, REG_READER, RISCV.AMO);
```

Comment:
From the comment below, I understand we should not
<img width="719" alt="Screenshot 2022-07-23 at 14 04 47" src="https://user-images.githubusercontent.com/11671435/180604120-74d363b3-2edf-4f99-8eee-5ce845137a8d.png">

Also, the RMW is used to guarantee that a load-store pair will behave atomically. In the case of AMO instructions, this is guaranteed differently (via using `memLoadValueExpr` and `memStoreValueExpr` instead of `memValueExpr`).

Reply:
But you need to distinguish between `Tag.RMW` and `RelRMW`. `Tag.RMW` is not used in `riscv.cat` at all IIRC.
So adding `Tag.RMW` is more about intention (i.e. an AMO is a RMW operation) rather than necessity (the tag is not used in the .cat).